File: research/families/f06_placement_fill_cif/audit/risk_set_expansion.py
"""Python contract for the native discrete competing-risk expander."""

from __future__ import annotations

from typing import Any

import numpy as np
import pandas as pd

SCHEMA_VERSION = "competing_risk_intervals.v1"
EVENT_CENSOR = 0
EVENT_FILL = 1
EVENT_ACK = 2


def expand_competing_risk_intervals_native(
    duration_ms: Any,
    event_kind: Any,
    bin_edges_ms: Any,
) -> pd.DataFrame:
    import narrowgate_cpp  # type: ignore

    if not hasattr(narrowgate_cpp, "expand_competing_risk_intervals"):
        raise RuntimeError("installed narrowgate_cpp lacks competing_risk_intervals.v1")
    output = narrowgate_cpp.expand_competing_risk_intervals(
        np.ascontiguousarray(duration_ms, dtype=np.float64),
        np.ascontiguousarray(event_kind, dtype=np.uint8),
        np.ascontiguousarray(bin_edges_ms, dtype=np.float64),
    )
    if str(output["schema_version"]) != SCHEMA_VERSION:
        raise RuntimeError("native risk-set schema identity changed")
    return pd.DataFrame(
        {name: np.asarray(values) for name, values in output.items() if name != "schema_version"}
    )


def expand_competing_risk_intervals_python(
    duration_ms: Any,
    event_kind: Any,
    bin_edges_ms: Any,
) -> pd.DataFrame:
    durations = np.asarray(duration_ms, dtype=float)
    events = np.asarray(event_kind, dtype=np.uint8)
    edges = np.asarray(bin_edges_ms, dtype=float)
    if len(durations) != len(events):
        raise ValueError("duration_ms and event_kind must align")
    if len(edges) < 2 or edges[0] != 0.0 or np.any(edges[1:] <= edges[:-1]):
        raise ValueError("bin edges must start at zero and increase")
    rows: list[dict[str, Any]] = []
    for row, (raw_duration, event) in enumerate(zip(durations, events, strict=True)):
        if not np.isfinite(raw_duration) or raw_duration < 0.0 or event > EVENT_ACK:
            raise ValueError("invalid duration/event")
        observed = min(float(raw_duration), float(edges[-1]))
        event_observed = event != EVENT_CENSOR and raw_duration <= edges[-1]
        event_bin = (
            int(np.searchsorted(edges[1:], raw_duration, side="left"))
            if event_observed
            else len(edges) - 1
        )
        for index, (start, end) in enumerate(zip(edges[:-1], edges[1:], strict=True)):
            exposure = max(0.0, min(observed, end) - start)
            is_event = event_observed and index == event_bin
            if exposure <= 0.0 and not is_event:
                break
            rows.append(
                {
                    "row_index": row,
                    "bin_index": index,
                    "interval_start_ms": start,
                    "interval_end_ms": end,
                    "exposure_fraction": np.clip(exposure / (end - start), 0.0, 1.0),
                    "fill_target": int(is_event and event == EVENT_FILL),
                    "ack_target": int(is_event and event == EVENT_ACK),
                }
            )
            if is_event or observed <= end:
                break
    return pd.DataFrame(rows)

File: research/families/f06_placement_fill_cif/cpp/risk_set_expansion.cpp
#include "risk_set_expansion.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace narrowgate_cpp {

RiskSetExpansionResult expand_competing_risk_intervals(
    ArrayView<double> duration_ms,
    ArrayView<std::uint8_t> event_kind,
    ArrayView<double> bin_edges_ms
) {
    if (duration_ms.size() != event_kind.size()) {
        throw std::invalid_argument("duration_ms and event_kind must align");
    }
    if (bin_edges_ms.size() < 2 || bin_edges_ms.front() != 0.0) {
        throw std::invalid_argument("bin_edges_ms must start at zero and contain two edges");
    }
    if (!std::is_sorted(bin_edges_ms.begin(), bin_edges_ms.end())) {
        throw std::invalid_argument("bin_edges_ms must be sorted ascending");
    }
    for (std::size_t i = 1; i < bin_edges_ms.size(); ++i) {
        if (!(bin_edges_ms[i] > bin_edges_ms[i - 1])) {
            throw std::invalid_argument("bin_edges_ms must be strictly increasing");
        }
    }

    RiskSetExpansionResult out;
    const std::size_t bins = bin_edges_ms.size() - 1;
    const double support_end = bin_edges_ms.back();
    for (std::size_t row = 0; row < duration_ms.size(); ++row) {
        const double raw_duration = duration_ms[row];
        const std::uint8_t kind = event_kind[row];
        if (!std::isfinite(raw_duration) || raw_duration < 0.0 || kind > 2) {
            throw std::invalid_argument("risk-set duration/event kind is invalid");
        }
        const double observed_duration = std::min(raw_duration, support_end);
        const bool event_observed = kind != 0 && raw_duration <= support_end;
        std::size_t event_bin = bins;
        if (event_observed) {
            const auto position = std::lower_bound(
                bin_edges_ms.begin() + 1,
                bin_edges_ms.end(),
                raw_duration
            );
            event_bin = static_cast<std::size_t>(position - bin_edges_ms.begin() - 1);
        }
        for (std::size_t bin = 0; bin < bins; ++bin) {
            const double start = bin_edges_ms[bin];
            const double end = bin_edges_ms[bin + 1];
            const double exposure = std::max(
                0.0, std::min(observed_duration, end) - start
            );
            const bool is_event_bin = event_observed && bin == event_bin;
            if (!(exposure > 0.0) && !is_event_bin) {
                break;
            }
            out.row_index.push_back(static_cast<std::int64_t>(row));
            out.bin_index.push_back(static_cast<std::int64_t>(bin));
            out.interval_start_ms.push_back(start);
            out.interval_end_ms.push_back(end);
            out.exposure_fraction.push_back(
                std::clamp(exposure / (end - start), 0.0, 1.0)
            );
            out.fill_target.push_back(
                static_cast<std::uint8_t>(is_event_bin && kind == 1)
            );
            out.ack_target.push_back(
                static_cast<std::uint8_t>(is_event_bin && kind == 2)
            );
            if (is_event_bin || observed_duration <= end) {
                break;
            }
        }
    }
    return out;
}

}  // namespace narrowgate_cpp
